internal/http/tmdb/search.go

```diff
@@ -0,0 +1,323 @@
+		return entry.FirstAirDate
+	}
+
+	return entry.ReleaseDate
+}
+
+func (date *Date) UnmarshalJSON(dateBytes []byte) error {
+	trimmedDateString := string(dateBytes[1 : len(dateBytes)-1])
+	parsed, err := time.Parse(time.DateOnly, trimmedDateString)
+	if err != nil {
+		return fmt.Errorf("cannot unmarshal Date due to error: %s", err.Error())
+	}
+
+	*date = Date{parsed}
+	return nil
+}
+
+// filterResultsInPlace will filter the given array of results IN PLACE by modifying
+// the provided slice and returning
+func filterResultsInPlace(results *[]SearchResultItem, metadata *media.FileMediaMetadata, filterFn func(dateFromResult time.Time, dateFromMetadata time.Time) bool) {
+	timeFromYear := func(year int) time.Time {
+		return time.Date(year, 1, 1, 0, 0, 0, 0, time.UTC)
+	}
+
+	yearFromMetadata := timeFromYear(*metadata.Year)
+	insertionIndex := 0
+	for _, v := range *results {
+		yearFromResult := timeFromYear(v.effectiveDate().Year())
+		if filterFn(yearFromResult, yearFromMetadata) {
+			(*results)[insertionIndex] = v
+			insertionIndex++
+		}
+	}
+
+	*results = (*results)[:insertionIndex]
+}
+
+func httpGetJsonResponse(urlPath string, targetInterface interface{}) error {
+	fmt.Printf("GET -> %s\n", urlPath)
+	resp, err := http.Get(urlPath)
+	if err != nil {
+		return &UnknownRequestError{fmt.Sprintf("failed to perform GET(%s) to TMDB: %s", urlPath, err.Error())}
+	}
+
+	defer resp.Body.Close()
+	respBody, err := io.ReadAll(resp.Body)
+
+	if resp.StatusCode != http.StatusOK {
+		var tmdbError tmdbError
+		if err := json.Unmarshal(respBody, &tmdbError); err != nil {
+			return &FailedRequestError{httpCode: resp.StatusCode, message: "non-OK response could not be unmarshalled", tmdbCode: -1}
+		}
+
+		return &FailedRequestError{httpCode: resp.StatusCode, message: tmdbError.StatusMessage, tmdbCode: tmdbError.StatusCode}
+	}
+
+	if err != nil {
+		return &UnknownRequestError{fmt.Sprintf("failed to read response body: %s", err.Error())}
+	}
+
+	if err := json.Unmarshal(respBody, targetInterface); err != nil {
+		return &UnknownRequestError{fmt.Sprintf("response JSON could not be unmarshalled: %s", err.Error())}
+	}
+
+	return nil
+}
+
+type (
+	tmdbError struct {
+		StatusCode    int    `json:"status_code"`
+		StatusMessage string `json:"status_message"`
+	}
+	FailedRequestError struct {
+		httpCode int
+		tmdbCode int
+		message  string
+	}
+	NoResultError       struct{}
+	MultipleResultError struct{ results []SearchResultItem }
+	UnknownRequestError struct{ reason string }
+	IllegalRequestError struct{ reason string }
+)
+
+func (err *UnknownRequestError) Error() string {
+	return fmt.Sprintf("unknown error occurred while communicating with TMDB: %s", err.reason)
+}
+func (err *IllegalRequestError) Error() string {
+	return fmt.Sprintf("illegal search request because %s", err.reason)
+}
+func (err *FailedRequestError) Error() string {
+	return fmt.Sprintf("Request failure (HTTP %d): %s", err.httpCode, err.message)
+}
```

Comment:
Doing errors like this is perfectly correct, but I would caution that it is probably overengineered.

Consider the place you are using this error:

```
return &FailedRequestError{httpCode: resp.StatusCode, message: "non-OK response could not be unmarshalled", tmdbCode: -1}
```

Unless you are using `errors.Is` or `errors.As` to handle specific errors in a specific way (or you are writing a public package), then it can be much simpler and more readable to just construct the error in place:

```
return fmt.Errorf("Request failure (HTTP %d): non-OK response could not be unmarshalled", resp.StatusCode)
```

Reply:
Yeah, so I do plan to actually capture and extract information out of these errors and provide them inside of some of the REST endpoint responses. For example, `MultipleResultError` contains a list of the results which we couldn't decide on, allowing the user to decide manually using the web frontend.

Given this reasoning, do errors like this seem like a good choice to you?